spec/Appendix A -- Persisted Documents.md

```diff
@@ -0,0 +1,262 @@
+# A. Appendix: Persisted Documents
+
+This appendix defines an optional extension to the GraphQL-over-HTTP protocol
+that allows for the usage of "persisted documents".
+
+:: A _persisted document_ is a GraphQL document (strictly: an
+[`ExecutableDocument`](https://spec.graphql.org/draft/#ExecutableDocument)) that
+has been persisted such that the server may retrieve it based on an identifier
+indicated in the HTTP request.
+
+This feature can be used as an operation allow-list, as a way of improving the
+caching of GraphQL operations, or just as a way of reducing the bandwidth
+consumed from sending the full GraphQL Document to the server on each request.
+
+Typically, support for the _persisted document_ feature is implemented via a
+"middleware" that sits in front of the GraphQL service and transforms a
+_persisted document request_ into a _GraphQL-over-HTTP request_.
+
+:: A _persisted operation_ is a _persisted document_ which contains only one
+GraphQL operation and all the fragments this operation references (recursively).
+
+## Identifying a Document
+
+:: A _document identifier_ is a string-based identifier that uniquely identifies
+a GraphQL Document.
+
+Note: A _document identifier_ must be unique, otherwise there is a risk of
+responses confusing the client. Even if the selection sets are identical, even
+whitespace changes may change the location from which errors are raised, and
+thus should generate different document identifiers.
+
+A _document identifier_ must either be a _prefixed document identifier_ or a
+_custom document identifier_.
+
+### Prefixed Document Identifier
+
+:: A _prefixed document identifier_ is a document identifier that contains at
+least one colon symbol (`:`). The text before the first colon symbol is called
+the {prefix}, and the text after it is called the {payload}. The {prefix}
+identifies the method of identification used. Applications may use their own
+identification methods by ensuring that the prefix starts `x-`; otherwise, all
+prefixes are reserved for reasons of future expansion.
+
+### SHA256 Hex Document Identifier
```

Comment:
I wonder if actual encryption is needed? what's the benefit of having `sha256` over any other opaque string? 🤔 
We can allow users how to encode/map their persisted documents, and then do matching based on that value?

Reply:
The aim of a specification like this is interoperability; so a client that supports persisted operations should be able to use a server that supports persisted operations without too much additional setup. Sharing details of the document identification method used out-of-band is supported (explicitly by `x-` prefix, or by custom identifiers); but for maximum compatibility there should be a shared baseline in my opinion.

Reply:
I expect that the benefit is mainly for APQ support, where a client can enable APQ without knowledge of the server implementation, and so a consistent implementation is essential.  SHA256 can be computed natively by modern browsers with a very low collision rate, making it a good choice in this scenario.

However, in a scenario where the identifiers are only known to the server, and must be registered with the server in order for the client to operate (as so far is documented in this RFC), they might as well be any opaque key returned by the query storage database.  If the database stores the queries with an auto-incrementing integer as an identifier, that would work just as well.

Even so, there are still some benefits to using a hash:
1. Should the query storage mechanism change, the identifiers will remain constant (but the same would be true of a GUID)
2. Inherit deduplication for stored queries
3. Enhanced security, as attackers can't easily guess valid identifiers, reducing the risk of unauthorized query execution

I'm fine with the current suggestion of `sha256:` but allowing for `x-id:` and similar.

Reply:
I see, thank you @Shane32 @benjie .
My point is not about the prefix or the method but for the need for encryption? a user can decide to use `operation-1` as the key (instead of an actual computed hash), and the result will be the same 🤔 🤔 🤔

Reply:
@dotansimha Indeed, that's already allowed under this spec. The issue is it requires coordination between server and client (they need to agree on how `operation-1` is derived). For maximum compatibility, if both server and client already know how to identify operations (e.g. standardized SHA256 hash) then all that's left is to transfer the docs from client to server, which can happen _after_ the client has been built, rather than before or during, and no configuration is required on server or client easing adoption for everyone and encouraging more people to use an operation allowlist.

Reply:
> However, in a scenario where the identifiers are only known to the server, and must be registered with the server in order for the client to operate (as so far is documented in this RFC)

I don't think that's what the spec says...

> A service which accepts a _persisted document request_ SHOULD support the
_SHA256 hex document identifier_ for compatibility.

^ So a client SHOULD be able to default to using the SHA256 hash, and will only need to be customized if that's not the case.

Reply:
Oh, right, but the queries still need to be transferred to the server is what you're saying. Yes, that's true, but it can be done _after_ the client is built (but before it's deployed). That's different from having to do it _during_ the build process; it means that clients can be built and persisted documents written even before the server _exists_. It also allows for arbitrary transfer of documents to the server (you can send them one-way on a pen drive through the mail if you want!).

Reply:
"Recommend" and "should" are the same according to RFC2119, so I think that is saying what's already there. If the server follows this recommendation, the client doesn't need any configuration. If the server _doesn't_ do this; _then_ you need to coordinate between client and server. For optimal interoperability, no coordination should be necessary.

Reply:
To be pedantic there, I'd argue that sharing the "sha256" method is still coordination between the client and server. Plus the documents need to be actually transferred (on a pen drive or avian carrier!). 

So coordination is always required but sha256 is a convenient and widespread default which we recommend (hence the should)?

Reply:
Indeed, it is a very light-touch asynchronous form of coordination. Essentially the coordination boils down to two things: 1. does the server support SHA256 hashes (don't necessarily need to ask the server this, it's a fact that should be established in the development team); 2. we need a way to send the operations+hashes to the server and to know when they have been persisted.

Client is informed that server supports SHA256 hashes.
Client: _performs build including generation of hashes_
Client (at some point later): _ships hashes and operations to server somehow_
Server (upon receipt): _stores queries+hashes somehow_
Client (after server has stored): _is deployed_

Since the client can't be deployed until the server has stored the queries/hashes, there is indeed coordination. The coordination is incredibly lightweight compared to alternatives where the server must generate hashes during the client build process.